File: src/main/java/com/projectswg/holocore/resources/support/global/commands/CombatCommand.java
/***********************************************************************************
 * Copyright (c) 2018 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is the first NGE emulator for Star Wars Galaxies founded on          *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create an emulator which will provide a server for players to    *
 * continue playing a game similar to the one they used to play. We are basing     *
 * it on the final publish of the game prior to end-game events.                   *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.resources.support.global.commands;

import com.projectswg.common.data.CRC;
import com.projectswg.common.data.combat.*;
import com.projectswg.holocore.resources.support.objects.swg.weapon.WeaponType;
import org.jetbrains.annotations.NotNull;

import java.util.EnumMap;
import java.util.Map;
import java.util.Objects;

public class CombatCommand {
	
	private final ValidTarget validTarget;
	private final boolean forceCombat;
	private final Map<WeaponType, String[]> animations;
	private final String [] defaultAnimation;
	private final AttackType attackType;
	private final double healthCost;
	private final double actionCost;
	private final double mindCost;
	private final double knockdownChance;
	private final boolean blinding;
	private final boolean bleeding;
	private final String triggerEffect;
	private final String triggerEffectHardpoint;
	private final DamageType damageType;
	private final DamageType elementalType;
	private final boolean ignoreDistance;
	private final boolean pvpOnly;
	private final int attackRolls;
	private final double percentAddFromWeapon;
	private final double bypassArmor;
	private final double hateDamageModifier;
	private final int hateAdd;
	private final int addedDamage;
	private final String buffNameTarget;
	private final String buffNameSelf;
	private final HitType hitType;
	private final String delayAttackEggTemplate;
	private final String delayAttackParticle;
	private final double initialDelayAttackInterval;
	private final double delayAttackInterval;
	private final int delayAttackLoops;
	private final DelayAttackEggPosition eggPosition;
	private final double coneLength;
	private final double coneWidth;
	private final HealAttrib healAttrib;
	private final String name;
	private final double maxRange;
	private final int crc;
	
	private CombatCommand(CombatCommandBuilder builder) {
		this.validTarget = builder.validTarget;
		this.forceCombat = builder.forceCombat;
		this.animations = builder.animations;
		this.defaultAnimation = Objects.requireNonNull(builder.defaultAnimation, "defaultAnimation");
		this.attackType = builder.attackType;
		this.healthCost = builder.healthCost;
		this.actionCost = builder.actionCost;
		this.mindCost = builder.mindCost;
		this.knockdownChance = builder.knockdownChance;
		this.blinding = builder.blinding;
		this.bleeding = builder.bleeding;
		this.triggerEffect = builder.triggerEffect;
		this.triggerEffectHardpoint = builder.triggerEffectHardpoint;
		this.damageType = builder.damageType;
		this.elementalType = builder.elementalType;
		this.ignoreDistance = builder.ignoreDistance;
		this.pvpOnly = builder.pvpOnly;
		this.attackRolls = builder.attackRolls;
		this.percentAddFromWeapon = builder.percentAddFromWeapon;
		this.bypassArmor = builder.bypassArmor;
		this.hateDamageModifier = builder.hateDamageModifier;
		this.hateAdd = builder.hateAdd;
		this.addedDamage = builder.addedDamage;
		this.buffNameTarget = builder.buffNameTarget;
		this.buffNameSelf = builder.buffNameSelf;
		this.hitType = builder.hitType;
		this.delayAttackEggTemplate = builder.delayAttackEggTemplate;
		this.delayAttackParticle = builder.delayAttackParticle;
		this.initialDelayAttackInterval = builder.initialDelayAttackInterval;
		this.delayAttackInterval = builder.delayAttackInterval;
		this.delayAttackLoops = builder.delayAttackLoops;
		this.eggPosition = builder.eggPosition;
		this.coneLength = builder.coneLength;
		this.coneWidth = builder.coneWidth;
		this.healAttrib = builder.healAttrib;
		this.name = builder.name;
		this.maxRange = builder.maxRange;
		this.crc = CRC.getCrc(this.name);
	}
	
	public ValidTarget getValidTarget() {
		return validTarget;
	}
	
	public boolean isForceCombat() {
		return forceCombat;
	}
	
	@NotNull
	public String[] getAnimations(WeaponType type) {
		return animations.getOrDefault(type, new String[0]);
	}
	
	@NotNull
	public String getRandomAnimation(WeaponType type) {
		String [] animations = this.animations.get(type);
		if (animations == null || animations.length == 0)
			animations = defaultAnimation;
		if (animations == null || animations.length == 0)
			return "";
		return animations[(int) (Math.random() * animations.length)];
	}
	
	@NotNull
	public String[] getDefaultAnimation() {
		return defaultAnimation;
	}
	
	public AttackType getAttackType() {
		return attackType;
	}
	
	public double getHealthCost() {
		return healthCost;
	}
	
	public double getActionCost() {
		return actionCost;
	}
	
	public double getMindCost() {
		return mindCost;
	}
	
	public double getKnockdownChance() {
		return knockdownChance;
	}
	
	public boolean isBlinding() {
		return blinding;
	}
	
	public boolean isBleeding() {
		return bleeding;
	}
	
	public String getTriggerEffect() {
		return triggerEffect;
	}
	
	public String getTriggerEffectHardpoint() {
		return triggerEffectHardpoint;
	}
	
	public DamageType getDamageType() {
		return damageType;
	}
	
	public DamageType getElementalType() {
		return elementalType;
	}
	
	public boolean isIgnoreDistance() {
		return ignoreDistance;
	}
	
	public boolean isPvpOnly() {
		return pvpOnly;
	}
	
	public int getAttackRolls() {
		return attackRolls;
	}
	
	public double getPercentAddFromWeapon() {
		return percentAddFromWeapon;
	}
	
	public double getBypassArmor() {
		return bypassArmor;
	}
	
	public double getHateDamageModifier() {
		return hateDamageModifier;
	}
	
	public int getHateAdd() {
		return hateAdd;
	}
	
	public int getAddedDamage() {
		return addedDamage;
	}
	
	public String getBuffNameTarget() {
		return buffNameTarget;
	}
	
	public String getBuffNameSelf() {
		return buffNameSelf;
	}
	
	public HitType getHitType() {
		return hitType;
	}
	
	public String getDelayAttackEggTemplate() {
		return delayAttackEggTemplate;
	}
	
	public String getDelayAttackParticle() {
		return delayAttackParticle;
	}
	
	public double getInitialDelayAttackInterval() {
		return initialDelayAttackInterval;
	}
	
	public double getDelayAttackInterval() {
		return delayAttackInterval;
	}
	
	public int getDelayAttackLoops() {
		return delayAttackLoops;
	}
	
	public DelayAttackEggPosition getEggPosition() {
		return eggPosition;
	}
	
	public double getConeLength() {
		return coneLength;
	}
	
	public double getConeWidth() {
		return coneWidth;
	}
	
	public HealAttrib getHealAttrib() {
		return healAttrib;
	}
	
	public String getName() {
		return name;
	}
	
	public static CombatCommandBuilder builder() {
		return new CombatCommandBuilder();
	}

	public double getMaxRange() {
		return maxRange;
	}

	public int getCrc() {
		return crc;
	}

	public static class CombatCommandBuilder {
		
		private final Map<WeaponType, String[]> animations;
		
		private ValidTarget validTarget;
		private boolean forceCombat;
		private String [] defaultAnimation;
		private AttackType attackType;
		private double healthCost;
		private double actionCost;
		private double mindCost;
		private double knockdownChance;
		private boolean blinding;
		private boolean bleeding;
		private String triggerEffect;
		private String triggerEffectHardpoint;
		private DamageType damageType;
		private DamageType elementalType;
		private boolean ignoreDistance;
		private boolean pvpOnly;
		private int attackRolls;
		private double percentAddFromWeapon;
		private double bypassArmor;
		private double hateDamageModifier;
		private int hateAdd;
		private int addedDamage;
		private String buffNameTarget;
		private String buffNameSelf;
		private HitType hitType;
		private String delayAttackEggTemplate;
		private String delayAttackParticle;
		private double initialDelayAttackInterval;
		private double delayAttackInterval;
		private int delayAttackLoops;
		private DelayAttackEggPosition eggPosition;
		private double coneLength;
		private double coneWidth;
		private HealAttrib healAttrib;
		private String name;
		private double maxRange;
		
		private CombatCommandBuilder() {
			this.animations = new EnumMap<>(WeaponType.class);
		}
		
		public CombatCommandBuilder withValidTarget(ValidTarget validTarget) {
			this.validTarget = validTarget;
			return this;
		}
		
		public CombatCommandBuilder withForceCombat(boolean forceCombat) {
			this.forceCombat = forceCombat;
			return this;
		}
		
		public CombatCommandBuilder withAnimations(WeaponType type, String[] animations) {
			this.animations.put(type, animations);
			return this;
		}
		
		public CombatCommandBuilder withDefaultAnimation(String[] defaultAnimation) {
			this.defaultAnimation = defaultAnimation;
			return this;
		}
		
		public CombatCommandBuilder withAttackType(AttackType attackType) {
			this.attackType = attackType;
			return this;
		}
		
		public CombatCommandBuilder withHealthCost(double healthCost) {
			this.healthCost = healthCost;
			return this;
		}
		
		public CombatCommandBuilder withActionCost(double actionCost) {
			this.actionCost = actionCost;
			return this;
		}
		
		public CombatCommandBuilder withMindCost(double mindCost) {
			this.mindCost = mindCost;
			return this;
		}
		
		public CombatCommandBuilder withKnockdownChance(double knockdownChance) {
			this.knockdownChance = knockdownChance;
			return this;
		}
		
		public CombatCommandBuilder withBlinding(boolean blinding) {
			this.blinding = blinding;
			return this;
		}
		
		public CombatCommandBuilder withBleeding(boolean bleeding) {
			this.bleeding = bleeding;
			return this;
		}
		
		public CombatCommandBuilder withTriggerEffect(String triggerEffect) {
			this.triggerEffect = triggerEffect;
			return this;
		}
		
		public CombatCommandBuilder withTriggerEffectHardpoint(String triggerEffectHardpoint) {
			this.triggerEffectHardpoint = triggerEffectHardpoint;
			return this;
		}
		
		public CombatCommandBuilder withDamageType(DamageType damageType) {
			this.damageType = damageType;
			return this;
		}
		
		public CombatCommandBuilder withElementalType(DamageType elementalType) {
			this.elementalType = elementalType;
			return this;
		}
		
		public CombatCommandBuilder withIgnoreDistance(boolean ignoreDistance) {
			this.ignoreDistance = ignoreDistance;
			return this;
		}
		
		public CombatCommandBuilder withPvpOnly(boolean pvpOnly) {
			this.pvpOnly = pvpOnly;
			return this;
		}
		
		public CombatCommandBuilder withAttackRolls(int attackRolls) {
			this.attackRolls = attackRolls;
			return this;
		}
		
		public CombatCommandBuilder withPercentAddFromWeapon(double percentAddFromWeapon) {
			this.percentAddFromWeapon = percentAddFromWeapon;
			return this;
		}
		
		public CombatCommandBuilder withBypassArmor(double bypassArmor) {
			this.bypassArmor = bypassArmor;
			return this;
		}
		
		public CombatCommandBuilder withHateDamageModifier(double hateDamageModifier) {
			this.hateDamageModifier = hateDamageModifier;
			return this;
		}
		
		public CombatCommandBuilder withHateAdd(int hateAdd) {
			this.hateAdd = hateAdd;
			return this;
		}
		
		public CombatCommandBuilder withAddedDamage(int addedDamage) {
			this.addedDamage = addedDamage;
			return this;
		}
		
		public CombatCommandBuilder withBuffNameTarget(String buffNameTarget) {
			this.buffNameTarget = buffNameTarget;
			return this;
		}
		
		public CombatCommandBuilder withBuffNameSelf(String buffNameSelf) {
			this.buffNameSelf = buffNameSelf;
			return this;
		}
		
		public CombatCommandBuilder withHitType(HitType hitType) {
			this.hitType = hitType;
			return this;
		}
		
		public CombatCommandBuilder withDelayAttackEggTemplate(String delayAttackEggTemplate) {
			this.delayAttackEggTemplate = delayAttackEggTemplate;
			return this;
		}
		
		public CombatCommandBuilder withDelayAttackParticle(String delayAttackParticle) {
			this.delayAttackParticle = delayAttackParticle;
			return this;
		}
		
		public CombatCommandBuilder withInitialDelayAttackInterval(double initialDelayAttackInterval) {
			this.initialDelayAttackInterval = initialDelayAttackInterval;
			return this;
		}
		
		public CombatCommandBuilder withDelayAttackInterval(double delayAttackInterval) {
			this.delayAttackInterval = delayAttackInterval;
			return this;
		}
		
		public CombatCommandBuilder withDelayAttackLoops(int delayAttackLoops) {
			this.delayAttackLoops = delayAttackLoops;
			return this;
		}
		
		public CombatCommandBuilder withEggPosition(DelayAttackEggPosition eggPosition) {
			this.eggPosition = eggPosition;
			return this;
		}
		
		public CombatCommandBuilder withConeLength(double coneLength) {
			this.coneLength = coneLength;
			return this;
		}
		
		public CombatCommandBuilder withConeWidth(double coneWidth) {
			this.coneWidth = coneWidth;
			return this;
		}
		
		public CombatCommandBuilder withHealAttrib(HealAttrib healAttrib) {
			this.healAttrib = healAttrib;
			return this;
		}
		
		public CombatCommandBuilder withName(String name) {
			this.name = name;
			return this;
		}

		public CombatCommandBuilder withMaxRange(double maxRange) {
			this.maxRange = maxRange;
			return this;
		}
		
		public CombatCommand build() {
			return new CombatCommand(this);
		}
		
	}
}

File: src/main/java/com/projectswg/holocore/resources/support/global/commands/CombatCommand.kt
/***********************************************************************************
 * Copyright (c) 2024 /// Project SWG /// www.projectswg.com                       *
 *                                                                                 *
 * ProjectSWG is an emulation project for Star Wars Galaxies founded on            *
 * July 7th, 2011 after SOE announced the official shutdown of Star Wars Galaxies. *
 * Our goal is to create one or more emulators which will provide servers for      *
 * players to continue playing a game similar to the one they used to play.        *
 *                                                                                 *
 * This file is part of Holocore.                                                  *
 *                                                                                 *
 * --------------------------------------------------------------------------------*
 *                                                                                 *
 * Holocore is free software: you can redistribute it and/or modify                *
 * it under the terms of the GNU Affero General Public License as                  *
 * published by the Free Software Foundation, either version 3 of the              *
 * License, or (at your option) any later version.                                 *
 *                                                                                 *
 * Holocore is distributed in the hope that it will be useful,                     *
 * but WITHOUT ANY WARRANTY; without even the implied warranty of                  *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the                   *
 * GNU Affero General Public License for more details.                             *
 *                                                                                 *
 * You should have received a copy of the GNU Affero General Public License        *
 * along with Holocore.  If not, see <http://www.gnu.org/licenses/>.               *
 ***********************************************************************************/
package com.projectswg.holocore.resources.support.global.commands

import com.projectswg.common.data.CRC
import com.projectswg.common.data.combat.*
import com.projectswg.holocore.resources.support.objects.swg.weapon.WeaponType
import java.util.*

class CombatCommand private constructor(builder: CombatCommandBuilder) {

	val validTarget: ValidTarget = builder.validTarget
	val isForceCombat: Boolean = builder.forceCombat
	private val animations: Map<WeaponType, Array<String>> = builder.animations
	private val defaultAnimation: Array<String> = requireNotNull(builder.defaultAnimation)
	val attackType: AttackType = builder.attackType
	val healthCost: Double = builder.healthCost
	val actionCost: Double = builder.actionCost
	val mindCost: Double = builder.mindCost
	val forceCost: Double = builder.forceCost
	val forceCostModifier: Double = builder.forceCostModifier
	val knockdownChance: Double = builder.knockdownChance
	val isBlinding: Boolean = builder.blinding
	val isBleeding: Boolean = builder.bleeding
	val isDizzying: Boolean = builder.dizzying
	val isStunning: Boolean = builder.stunning
	val triggerEffect: String = builder.triggerEffect
	val triggerEffectHardpoint: String = builder.triggerEffectHardpoint
	val targetEffect: String = builder.targetEffect
	val targetEffectHardpoint: String = builder.targetEffectHardpoint
	val damageType: DamageType = builder.damageType
	val elementalType: DamageType = builder.elementalType
	val isIgnoreDistance: Boolean = builder.ignoreDistance
	val isPvpOnly: Boolean = builder.pvpOnly
	val percentAddFromWeapon: Double = builder.percentAddFromWeapon
	val bypassArmor: Double = builder.bypassArmor
	val hateDamageModifier: Double = builder.hateDamageModifier
	val hateAdd: Int = builder.hateAdd
	val addedDamage: Int = builder.addedDamage
	val buffNameTarget: String = builder.buffNameTarget
	val buffNameSelf: String = builder.buffNameSelf
	val hitType: HitType = builder.hitType
	val delayAttackEggTemplate: String = builder.delayAttackEggTemplate
	val delayAttackParticle: String = builder.delayAttackParticle
	val initialDelayAttackInterval: Double = builder.initialDelayAttackInterval
	val delayAttackInterval: Double = builder.delayAttackInterval
	val delayAttackLoops: Int = builder.delayAttackLoops
	val eggPosition: DelayAttackEggPosition = builder.eggPosition
	val coneLength: Double = builder.coneLength
	val coneWidth: Double = builder.coneWidth
	val healAttrib: HealAttrib = builder.healAttrib
	val name: String = requireNotNull(builder.name)
	val maxRange: Double = builder.maxRange
	val crc: Int = CRC.getCrc(builder.name)
	val hitLocation: HitLocation = builder.hitLocation

	fun getAnimations(type: WeaponType): Array<String> {
		return animations.getOrDefault(type, Array(0) { "" })
	}

	fun getRandomAnimation(type: WeaponType): String {
		var animations = animations[type]
		if (animations.isNullOrEmpty()) animations = defaultAnimation
		if (animations.isEmpty()) return ""
		return animations[(Math.random() * animations.size).toInt()]
	}

	class CombatCommandBuilder {
		val animations: MutableMap<WeaponType, Array<String>> = EnumMap(WeaponType::class.java)

		var validTarget: ValidTarget = ValidTarget.STANDARD
		var forceCombat: Boolean = false
		var defaultAnimation: Array<String> = Array(0) { "" }
		var attackType: AttackType = AttackType.SINGLE_TARGET
		var healthCost: Double = 0.0
		var actionCost: Double = 0.0
		var mindCost: Double = 0.0
		var forceCost: Double = 0.0
		var forceCostModifier: Double = 0.0
		var knockdownChance: Double = 0.0
		var blinding: Boolean = false
		var bleeding: Boolean = false
		var dizzying: Boolean = false
		var stunning: Boolean = false
		var triggerEffect: String = ""
		var triggerEffectHardpoint: String = ""
		var targetEffect: String = ""
		var targetEffectHardpoint: String = ""
		var damageType: DamageType = DamageType.KINETIC
		var elementalType: DamageType = DamageType.KINETIC
		var ignoreDistance: Boolean = false
		var pvpOnly: Boolean = false
		var percentAddFromWeapon: Double = 0.0
		var bypassArmor: Double = 0.0
		var hateDamageModifier: Double = 0.0
		var hateAdd: Int = 0
		var addedDamage: Int = 0
		var buffNameTarget: String = ""
		var buffNameSelf: String = ""
		var hitType: HitType = HitType.ATTACK
		var delayAttackEggTemplate: String = ""
		var delayAttackParticle: String = ""
		var initialDelayAttackInterval: Double = 0.0
		var delayAttackInterval: Double = 0.0
		var delayAttackLoops: Int = 0
		var eggPosition: DelayAttackEggPosition = DelayAttackEggPosition.TARGET
		var coneLength: Double = 0.0
		var coneWidth: Double = 0.0
		var healAttrib: HealAttrib = HealAttrib.HEALTH
		var name: String = ""
		var maxRange: Double = 0.0
		var hitLocation: HitLocation = HitLocation.HIT_LOCATION_BODY

		fun withValidTarget(validTarget: ValidTarget): CombatCommandBuilder {
			this.validTarget = validTarget
			return this
		}

		fun withForceCombat(forceCombat: Boolean): CombatCommandBuilder {
			this.forceCombat = forceCombat
			return this
		}

		fun withAnimations(type: WeaponType, animations: Array<String>): CombatCommandBuilder {
			this.animations[type] = animations
			return this
		}

		fun withDefaultAnimation(defaultAnimation: Array<String>): CombatCommandBuilder {
			this.defaultAnimation = defaultAnimation
			return this
		}

		fun withAttackType(attackType: AttackType): CombatCommandBuilder {
			this.attackType = attackType
			return this
		}

		fun withHealthCost(healthCost: Double): CombatCommandBuilder {
			this.healthCost = healthCost
			return this
		}

		fun withActionCost(actionCost: Double): CombatCommandBuilder {
			this.actionCost = actionCost
			return this
		}

		fun withMindCost(mindCost: Double): CombatCommandBuilder {
			this.mindCost = mindCost
			return this
		}

		fun withForceCost(forceCost: Double): CombatCommandBuilder {
			this.forceCost = forceCost
			return this
		}

		fun withForceCostModifier(forceCostModifier: Double): CombatCommandBuilder {
			this.forceCostModifier = forceCostModifier
			return this
		}

		fun withKnockdownChance(knockdownChance: Double): CombatCommandBuilder {
			this.knockdownChance = knockdownChance
			return this
		}

		fun withBlinding(blinding: Boolean): CombatCommandBuilder {
			this.blinding = blinding
			return this
		}

		fun withBleeding(bleeding: Boolean): CombatCommandBuilder {
			this.bleeding = bleeding
			return this
		}

		fun withDizzying(dizzying: Boolean): CombatCommandBuilder {
			this.dizzying = dizzying
			return this
		}

		fun withStunning(stunning: Boolean): CombatCommandBuilder {
			this.stunning = stunning
			return this
		}

		fun withTriggerEffect(triggerEffect: String): CombatCommandBuilder {
			this.triggerEffect = triggerEffect
			return this
		}

		fun withTriggerEffectHardpoint(triggerEffectHardpoint: String): CombatCommandBuilder {
			this.triggerEffectHardpoint = triggerEffectHardpoint
			return this
		}

		fun withTargetEffect(targetEffect: String): CombatCommandBuilder {
			this.targetEffect = targetEffect
			return this
		}

		fun withTargetEffectHardpoint(targetEffectHardpoint: String): CombatCommandBuilder {
			this.targetEffectHardpoint = targetEffectHardpoint
			return this
		}

		fun withDamageType(damageType: DamageType): CombatCommandBuilder {
			this.damageType = damageType
			return this
		}

		fun withElementalType(elementalType: DamageType): CombatCommandBuilder {
			this.elementalType = elementalType
			return this
		}

		fun withIgnoreDistance(ignoreDistance: Boolean): CombatCommandBuilder {
			this.ignoreDistance = ignoreDistance
			return this
		}

		fun withPvpOnly(pvpOnly: Boolean): CombatCommandBuilder {
			this.pvpOnly = pvpOnly
			return this
		}

		fun withPercentAddFromWeapon(percentAddFromWeapon: Double): CombatCommandBuilder {
			this.percentAddFromWeapon = percentAddFromWeapon
			return this
		}

		fun withBypassArmor(bypassArmor: Double): CombatCommandBuilder {
			this.bypassArmor = bypassArmor
			return this
		}

		fun withHateDamageModifier(hateDamageModifier: Double): CombatCommandBuilder {
			this.hateDamageModifier = hateDamageModifier
			return this
		}

		fun withHateAdd(hateAdd: Int): CombatCommandBuilder {
			this.hateAdd = hateAdd
			return this
		}

		fun withAddedDamage(addedDamage: Int): CombatCommandBuilder {
			this.addedDamage = addedDamage
			return this
		}

		fun withBuffNameTarget(buffNameTarget: String): CombatCommandBuilder {
			this.buffNameTarget = buffNameTarget
			return this
		}

		fun withBuffNameSelf(buffNameSelf: String): CombatCommandBuilder {
			this.buffNameSelf = buffNameSelf
			return this
		}

		fun withHitType(hitType: HitType): CombatCommandBuilder {
			this.hitType = hitType
			return this
		}

		fun withDelayAttackEggTemplate(delayAttackEggTemplate: String): CombatCommandBuilder {
			this.delayAttackEggTemplate = delayAttackEggTemplate
			return this
		}

		fun withDelayAttackParticle(delayAttackParticle: String): CombatCommandBuilder {
			this.delayAttackParticle = delayAttackParticle
			return this
		}

		fun withInitialDelayAttackInterval(initialDelayAttackInterval: Double): CombatCommandBuilder {
			this.initialDelayAttackInterval = initialDelayAttackInterval
			return this
		}

		fun withDelayAttackInterval(delayAttackInterval: Double): CombatCommandBuilder {
			this.delayAttackInterval = delayAttackInterval
			return this
		}

		fun withDelayAttackLoops(delayAttackLoops: Int): CombatCommandBuilder {
			this.delayAttackLoops = delayAttackLoops
			return this
		}

		fun withEggPosition(eggPosition: DelayAttackEggPosition): CombatCommandBuilder {
			this.eggPosition = eggPosition
			return this
		}

		fun withConeLength(coneLength: Double): CombatCommandBuilder {
			this.coneLength = coneLength
			return this
		}

		fun withConeWidth(coneWidth: Double): CombatCommandBuilder {
			this.coneWidth = coneWidth
			return this
		}

		fun withHealAttrib(healAttrib: HealAttrib): CombatCommandBuilder {
			this.healAttrib = healAttrib
			return this
		}

		fun withName(name: String): CombatCommandBuilder {
			this.name = name
			return this
		}

		fun withMaxRange(maxRange: Double): CombatCommandBuilder {
			this.maxRange = maxRange
			return this
		}

		fun withHitLocation(hitLocation: HitLocation): CombatCommandBuilder {
			this.hitLocation = hitLocation
			return this
		}

		fun build(): CombatCommand {
			return CombatCommand(this)
		}
	}

	companion object {
		fun builder(): CombatCommandBuilder {
			return CombatCommandBuilder()
		}
	}
}
